packages/amplify_api/lib/src/graphql/graphql_request_factory.dart

```diff
@@ -0,0 +1,160 @@
+
+  static GraphQLRequestFactory get instance => _instance;
+
+  String _getModelType(ModelFieldTypeEnum val) {
+    switch (val) {
+      case ModelFieldTypeEnum.string:
+        return "String";
+      case ModelFieldTypeEnum.int:
+        return "Int";
+      case ModelFieldTypeEnum.model:
+        return "ID";
+      default:
+        return "Error: unsupported type!";
+    }
+  }
+
+  String _getFieldsFromModelType(ModelSchema schema) {
+    // schema has been validated & schema.fields is non-nullable
+    Map<String, ModelField> fieldsMap = schema.fields!;
+    return fieldsMap.entries
+        .map((entry) => entry.value.association == null ? entry.key : '')
+        .toList()
+        .join(' ');
+  }
+
+  ModelSchema _getAndValidateSchema(
+      ModelType modelType, GraphQLRequestOperation operation) {
+    ModelProviderInterface? provider = AmplifyAPI.instance.getModelProvider();
+
+    if (provider == null) {
+      throw ApiException('No modelProvider found',
+          recoverySuggestion:
+              'Pass in a modelProvider instance while instantiating APIPlugin');
+    }
+
+    ModelSchema schema = provider.modelSchemas.firstWhere(
+        (elem) => elem.name == modelType.modelName(),
+        orElse: () => throw ApiException(
+            'No schema found for the ModelType provided',
+            recoverySuggestion:
+                'Pass in a valid modelProvider instance while instantiating APIPlugin or provide a valid ModelType'));
+
+    if (schema.fields == null) {
+      throw ApiException('Schema found does not have a fields property',
+          recoverySuggestion:
+              'Pass in a valid modelProvider instance while instantiating APIPlugin');
+    }
+
+    if (operation == GraphQLRequestOperation.list &&
+        schema.pluralName == null) {
+      throw ApiException('No schema name found',
+          recoverySuggestion:
+              'Pass in a valid modelProvider instance while instantiating APIPlugin or provide a valid ModelType');
+    }
+
+    return schema;
+  }
+
+  DocumentInputs _buildDocumentInputs(
+      ModelSchema schema, GraphQLRequestOperation operation) {
+    String upperOutput = '';
+    String lowerOutput = '';
+    List<String> upperList = [];
+    List<String> lowerList = [];
+
+    // build inputs based on request operation
+    switch (operation) {
+      case GraphQLRequestOperation.get:
+        upperList.add(r"$id: ID!");
+        lowerList.add(r"id: $id");
+        lowerOutput = "(${lowerList.join(", ")})";
+        break;
+      case GraphQLRequestOperation.list:
+        break;
+      default:
+        schema.fields!.forEach((field, val) {
+          upperList.add("\$$field: ${_getModelType(val.type.fieldType)}");
+          lowerList.add("$field: \$$field");
+        });
+        lowerOutput = "(input: { ${lowerList.join(", ")} })";
+    }
+
+    if (upperList.length == 0 || lowerList.length == 0) {
+      return DocumentInputs("", "");
```

Comment:
This case will need to be handled in a slightly different way below, since selection sets require 1 or more values:
- https://spec.graphql.org/June2018/#sec-Selection-Sets
- https://spec.graphql.org/June2018/#sec-Grammar-Notation

So, the following is invalid GraphQL:

```graphql
query listItems {
  listItems { } 
}
```

Now, I'm not sure in what scenario this condition would ever hold true, since there will never be a model with no fields. Just thought I'd mention it, though - no need to fix right now.

Reply:
Thanks for the feedback, sounds like I can clean up the list operations a little more. Currently how it sets up actually produces this document: 

```graphql
query listBlogs {
   listBlogs {
    items {
      id
      name
      createdAt
    }
    nextToken
  } 
}
```

Fields (id, name, createdAt) are added in another step. That being said, your comment highlights this can probably be consolidated so it's easier to follow/read. Will address in the next iteration 👍